src/v/iceberg/rest_client/catalog_client.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "http/client.h"
+#include "http/request_builder.h"
+#include "iceberg/rest_client/retry_policy.h"
+#include "iceberg/rest_client/types.h"
+#include "json/document.h"
+#include "utils/named_type.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+
+namespace iceberg::rest_client {
+
+using base_path = named_type<ss::sstring, struct base_path_t>;
+using prefix_path = named_type<ss::sstring, struct prefix_t>;
+using api_version = named_type<ss::sstring, struct api_version_t>;
+
+// Holds parts of a root path used by catalog client
+struct path_components {
+    path_components(
+      std::optional<base_path> base = std::nullopt,
+      std::optional<prefix_path> prefix = std::nullopt,
+      std::optional<api_version> api_version = std::nullopt);
+
+    // Returns root path for use in API calls not related to oauth token, joins
+    // base, api version and prefix with slashes and adds a terminating slash.
+    ss::sstring root_path() const;
+    // Same as root path but does not include the prefix
+    ss::sstring token_api_path() const;
+
+private:
+    base_path _base;
+    prefix_path _prefix;
+    api_version _api_version;
+};
+
+// helper intended to catch JSON parse errors when parsing an iobuf to json
+// document before passing the document to more specific parsers
+expected<json::Document> parse_json(iobuf&& raw_response);
+
+// The catalog client enables high level operations against a catalog rest api.
+// It also ensures that a token is available before making calls. A client is
+// lightweight and expected to be created for single operations or closely
+// related operations.
```

Comment:
We definitely want to be using one http client within just one catalog client at a time. In the remote client pool the lease associated to a client helps with this. Since we do not have a pool right now, I think passing in a unique ptr to an http client in the ctor and having the catalog client own it might be better than the reference approach we have currently.

Reply:
I have added a `client_source` which currently hands out unique ptrs to http clients. The catalog client will acquire and own a unique http client per call, and will also shut it down in all code paths (a method is added for shutdown to the interface).

The client source is passed as a reference to the catalog client. Right now it can be a factory object creating new clients per call to `acquire`. Once we have a pool we can switch to using that as client source, which will also require passing out leases instead of `unique_ptr`